projects/smoke_mcp3427/src/main.c

```diff
@@ -0,0 +1,210 @@
+      .adc_data_trigger_event = MCP3427_DATA_TRIGGER_EVENT,
+  },
+  {
+      .port = I2C_PORT_2,
+      .addr_pin_0 = MCP3427_PIN_STATE_LOW,
+      .addr_pin_1 = MCP3427_PIN_STATE_HIGH,
+      .sample_rate = SMOKE_SAMPLE_RATE,
+      .amplifier_gain = SMOKE_AMP_GAIN,
+      .conversion_mode = SMOKE_CONVERSION_MODE,
+      .adc_data_ready_event = MCP3427_DATA_READY_EVENT,
+      .adc_data_trigger_event = MCP3427_DATA_TRIGGER_EVENT,
+  },
+  {
+      .port = I2C_PORT_2,
+      .addr_pin_0 = MCP3427_PIN_STATE_HIGH,
+      .addr_pin_1 = MCP3427_PIN_STATE_FLOAT,
+      .sample_rate = SMOKE_SAMPLE_RATE,
+      .amplifier_gain = SMOKE_AMP_GAIN,
+      .conversion_mode = SMOKE_CONVERSION_MODE,
+      .adc_data_ready_event = MCP3427_DATA_READY_EVENT,
+      .adc_data_trigger_event = MCP3427_DATA_TRIGGER_EVENT,
+  },
+  {
+      .port = I2C_PORT_2,
+      .addr_pin_0 = MCP3427_PIN_STATE_HIGH,
+      .addr_pin_1 = MCP3427_PIN_STATE_HIGH,
+      .sample_rate = SMOKE_SAMPLE_RATE,
+      .amplifier_gain = SMOKE_AMP_GAIN,
+      .conversion_mode = SMOKE_CONVERSION_MODE,
+      .adc_data_ready_event = MCP3427_DATA_READY_EVENT,
+      .adc_data_trigger_event = MCP3427_DATA_TRIGGER_EVENT,
+  },
+  {
+      .port = I2C_PORT_2,
+      .addr_pin_0 = MCP3427_PIN_STATE_FLOAT,
+      .addr_pin_1 = MCP3427_PIN_STATE_HIGH,
+      .sample_rate = SMOKE_SAMPLE_RATE,
+      .amplifier_gain = SMOKE_AMP_GAIN,
+      .conversion_mode = SMOKE_CONVERSION_MODE,
+      .adc_data_ready_event = MCP3427_DATA_READY_EVENT,
+      .adc_data_trigger_event = MCP3427_DATA_TRIGGER_EVENT,
+  },
+  // Current sense mcp3427 settings
+  {
+      .port = I2C_PORT_1,
+      .addr_pin_0 = MCP3427_PIN_STATE_LOW,
+      .addr_pin_1 = MCP3427_PIN_STATE_LOW,
+      .sample_rate = SMOKE_SAMPLE_RATE,
+      .amplifier_gain = SMOKE_AMP_GAIN,
+      .conversion_mode = SMOKE_CONVERSION_MODE,
+      .adc_data_ready_event = MCP3427_DATA_READY_EVENT,
+      .adc_data_trigger_event = MCP3427_DATA_TRIGGER_EVENT,
+  }
+};
+
+static void prv_mcp3427_callback(int16_t value_ch1, int16_t value_ch2, void *context) {
+  SmokeMcp3427Data *data = context;
+  int16_t value = (SENSE_MCP3427_CHANNEL == MCP3427_CHANNEL_1) ? value_ch1 : value_ch2;
+  if (data->mcp3427_data_point < SMOKE_DATA_POINT_CURRENT) {
+    LOG_DEBUG("Voltage sense mcp3427 ID = %d; Value = %d\n", data->mcp3427_data_point, value);
+  } else {
+    LOG_DEBUG("Current sense mcp3427 ID = %d; Value = %d\n", data->mcp3427_data_point, value);
+  }
+}
+
+static void prv_mcp3527_fault_callback(void *context) {
+  SmokeMcp3427Data *data = context;
+
+  if (data->mcp3427_data_point < SMOKE_DATA_POINT_CURRENT) {
+    LOG_WARN("Voltage sense mcp3427 ID = %d encountered too many faults\n",
+             data->mcp3427_data_point);
+  } else {
+    LOG_WARN("Current sense mcp3427 ID = %d encountered too many faults", data->mcp3427_data_point);
+  }
+}
+
+int main() {
+  soft_timer_init();
+  event_queue_init();
+  interrupt_init();
+  gpio_init();
```

Comment:
`interrupt_init()` has to go before `soft_timer_init()` since `soft_timer` depends on `interrupt`.